presto-main/src/main/java/com/facebook/presto/sql/analyzer/MaterializedViewQueryOptimizer.java

```diff
@@ -71,418 +86,748 @@
     {
-        // TODO: Implement ways to handle non-optimizable query without throw/catch. https://github.com/prestodb/presto/issues/16541
-        try {
-            MaterializedViewInformationExtractor materializedViewInformationExtractor = new MaterializedViewInformationExtractor();
-            materializedViewInformationExtractor.process(materializedViewQuery);
-            materializedViewInfo = materializedViewInformationExtractor.getMaterializedViewInfo();
-            return process(node);
-        }
-        catch (Exception ex) {
-            logger.warn("Failed to rewrite query with materialized view with following exception: %s", ex.getMessage());
+        if (!node.getFrom().isPresent()) {
+            log.debug("Visited QuerySpecification had no from field: \n%s", node);
             return node;
         }
+        // If a query specification has a Table as a source, it can potentially be rewritten,
+        // so hand control over to LeafSubqueryVisitor via rewriteQuerySpecificationIfCompatible
+        Relation from = node.getFrom().get();
+        if (from instanceof Table) {
+            return rewriteQuerySpecificationIfCompatible(node, (Table) from);
+        }
+
+        if (from instanceof AliasedRelation
+                && ((AliasedRelation) from).getRelation() instanceof Table) {
+            AliasedRelation aliasedRelation = (AliasedRelation) from;
+            return rewriteQuerySpecificationIfCompatible(node, (Table) aliasedRelation.getRelation());
+        }
+
+        Relation newFrom = processSameType(from);
+        if (from == newFrom) {
+            return node;
+        }
+
+        if (node.getLocation().isPresent()) {
+            return new QuerySpecification(
+                    node.getLocation().get(),
+                    node.getSelect(),
+                    Optional.of(newFrom),
+                    node.getWhere(),
+                    node.getGroupBy(),
+                    node.getHaving(),
+                    node.getOrderBy(),
+                    node.getOffset(),
+                    node.getLimit());
+        }
+
+        return new QuerySpecification(
+                node.getSelect(),
+                Optional.of(newFrom),
+                node.getWhere(),
+                node.getGroupBy(),
+                node.getHaving(),
+                node.getOrderBy(),
+                node.getOffset(),
+                node.getLimit());
     }
 
     @Override
     protected Node visitNode(Node node, Void context)
     {
-        for (Node child : node.getChildren()) {
-            process(child, context);
-        }
         return node;
     }
 
     @Override
     protected Node visitQuery(Query node, Void context)
     {
+        QueryBody newQueryBody = processSameType(node.getQueryBody());
+        Optional<With> newWith = node.getWith().map(this::processSameType);
+        boolean withSame = !node.getWith().isPresent() || (node.getWith().get()) == newWith.get();
```

Comment:
need to check if `newWith` is present before calling `get()`

Reply:
Do we? The only way that `newWith` is not present is if:

1. `Optional.map` does not follow it's contract of returning an empty `Optional` if the calling `Optional` was empty
2. `node.getWith()` returns something different between these two calls

To me it seems like neither of these is possible, I consider the `node.getWith().isPresent()` check as good as one to `newWith.isPresent()`, and way less verbose.